src/bootstrap/lib.rs

```diff
@@ -65,9 +65,15 @@
 //! also check out the `src/bootstrap/README.md` file for more information.
 
 #![deny(warnings)]
+#![feature(associated_consts)]
+#![feature(core_intrinsics)]
```

Comment:
Yeah, I think it's a possible answer to replace the current HashMap impl with `Map<TypeId, Any>` (where `Any` is `HashMap<StepImplementor, StepImplementor::Output>`) with macros which generate the necessary relationships. Slightly annoying to have to write out all the steps which could get cached, but certainly possible.